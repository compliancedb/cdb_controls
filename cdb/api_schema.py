class ApiSchema:
    @staticmethod
    def url_for_owner_projects(host, project_data):
        return host + '/api/v1/projects/' + project_data["owner"] + "/"

    @staticmethod
    def url_for_project(host, project_data):
        return ApiSchema.url_for_owner_projects(host, project_data) + project_data["name"]

    @staticmethod
    def url_for_releases(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/releases/'

    @staticmethod
    def url_for_commit(host, project_data, commit):
        return ApiSchema.url_for_project(host, project_data) + '/commits/' + commit

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_artifact(host, project_data, sha256_digest):
        return ApiSchema.url_for_artifacts(host, project_data) + sha256_digest

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_release(host, project_data, release_number):
        return ApiSchema.url_for_releases(host, project_data) + release_number