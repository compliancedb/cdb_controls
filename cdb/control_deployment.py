def control_deployment_approved(approvals):
    for approval in approvals:
        if approval["state"] == "APPROVED":
            return True
    return False
