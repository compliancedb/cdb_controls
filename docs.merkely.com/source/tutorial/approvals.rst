Creating Approvals
==================

Many software development processes require an active decision to document an approval for deployment. This Human-in-the-Loop decision must be taken by someone responsible for accepting the risk of a given deployment. In order to be able to accept the risk, two criteria must be fulfilled:

1. The change must be understood
2. The acceptance must be documented

You can create the documentation for this approval automatically in your pipelines using Merkely's :ref:`log_approval` command.

.. image:: ../images/approvals.png

Merkely makes a dashboard for the cumulative change in the deployment, based on a list of the source git commits.  Provide two git references: one for what is currently in production, and one for what you would like to deploy.  The command will generate the git commit list between these two commits.

In this example from the image above, to create the approval for deploying the artifact created at the master commit you would use the following parameters in the :ref:`log_approval`:

+------------------------------+--------------------+
| MERKELY_OLDEST_SRC_COMMITISH | production         |
+------------------------------+--------------------+
| MERKELY_NEWEST_SRC_COMMITISH | master             |
+------------------------------+--------------------+

Pipeline approvals vs. Merkely approvals
----------------------------------------

Merkely provides two methods for documenting deployment approvals:

* Pipeline approval: an approval accepted externally to Merkely, for instance in an external CI tool.
* Merkely approval: an approval to be accepted in Merkely.

To create a pipeline approval, simply set the :code:`MERKELY_IS_APPROVED` parameter to :code:`TRUE`.  This will allow you to document an approval for example made in your CI tool and set it immediately to approved.  If you set this instead to :code:`FALSE`, it is possible to approve in the Merkely application.

Tracking branches for production
--------------------------------
In a typical setup, you need to add a :code:`production` tracking branch in git that is updated on every deployment. For example:

.. code-block:: bash
    :linenos:

    git checkout production
    git merge --ff-only master
    # YOUR DEPLOYMENT COMMAND HERE
    git push origin production



