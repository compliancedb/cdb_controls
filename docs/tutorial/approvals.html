<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Creating Approvals &#8212; Merkely  documentation</title>
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/merkely-custom.css" />
    <link rel="stylesheet" type="text/css" href="../_static/merkely-custom.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="&lt;no title&gt;" href="../concepts/index.html" />
    <link rel="prev" title="Logging Deployments" href="deployments.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html"><span><img src="../_static/w-merkely-white.png"></span>
           </a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Contents <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><p class="caption"><span class="caption-text">Quick Start:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="pipefile.html">Defining Your Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="binary_provenance.html">Logging Artifacts</a></li>
<li class="toctree-l1"><a class="reference internal" href="controls.html">Logging Controls</a></li>
<li class="toctree-l1"><a class="reference internal" href="deployments.html">Logging Deployments</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Creating Approvals</a></li>
</ul>
<p class="caption"><span class="caption-text">Concepts:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../concepts/devops_change_is_the_new_normal.html">DevOps: Change is the New Normal</a></li>
<li class="toctree-l1"><a class="reference internal" href="../concepts/devops_and_change_management.html">DevOps and Change Management</a></li>
<li class="toctree-l1"><a class="reference internal" href="../concepts/value_stream.html">Mapping your value stream</a></li>
<li class="toctree-l1"><a class="reference internal" href="../concepts/binary_provenance.html">Understanding Binary Provenance</a></li>
</ul>
<p class="caption"><span class="caption-text">Command Reference:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference/control_deployment.html">Control Deployment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/declare_pipeline.html">Declare Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/log_approval.html">Log Approval</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/log_artifact.html">Log Artifact</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/log_deployment.html">Log Deployment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/log_evidence.html">Log Evidence</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/log_test.html">Log Test</a></li>
</ul>
<p class="caption"><span class="caption-text">Fingerprints:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../fingerprints/docker_fingerprint.html">Docker Fingerprint</a></li>
<li class="toctree-l1"><a class="reference internal" href="../fingerprints/file_fingerprint.html">File Fingerprint</a></li>
<li class="toctree-l1"><a class="reference internal" href="../fingerprints/sha256_fingerprint.html">Sha256 Fingerprint</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="creating-approvals">
<h1>Creating Approvals<a class="headerlink" href="#creating-approvals" title="Permalink to this headline">¶</a></h1>
<p>Many software development processes require an active decision to document an approval for deployment. This Human-in-the-Loop decision must be taken by someone responsible for accepting the risk of a given deployment. In order to be able to accept the risk, two criteria must be fulfilled:</p>
<ol class="arabic simple">
<li><p>The change must be understood</p></li>
<li><p>The acceptance must be documented</p></li>
</ol>
<p>You can create the documentation for this approval automatically in your pipelines using Merkely’s <span class="xref std std-ref">log_approval</span> command.</p>
<img alt="../_images/approvals.png" src="../_images/approvals.png" />
<p>Merkely makes a dashboard for the cumulative change in the deployment, based on a list of the source git commits.  Provide two git references: one for what is currently in production, and one for what you would like to deploy.  The command will generate the git commit list between these two commits.</p>
<p>In this example from the image above, to create the approval for deploying the artifact created at the master commit you would use the following parameters in the <span class="xref std std-ref">log_approval</span>:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 60%" />
<col style="width: 40%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>MERKELY_OLDEST_SRC_COMMITISH</p></td>
<td><p>production</p></td>
</tr>
<tr class="row-even"><td><p>MERKELY_NEWEST_SRC_COMMITISH</p></td>
<td><p>master</p></td>
</tr>
</tbody>
</table>
<div class="section" id="pipeline-approvals-vs-merkely-approvals">
<h2>Pipeline approvals vs. Merkely approvals<a class="headerlink" href="#pipeline-approvals-vs-merkely-approvals" title="Permalink to this headline">¶</a></h2>
<p>Merkely provides two methods for documenting deployment approvals:</p>
<ul class="simple">
<li><p>Pipeline approval: an approval accepted externally to Merkely, for instance in an external CI tool.</p></li>
<li><p>Merkely approval: an approval to be accepted in Merkely.</p></li>
</ul>
<p>To create a pipeline approval, simply set the <code class="code docutils literal notranslate"><span class="pre">MERKELY_IS_APPROVED</span></code> parameter to <code class="code docutils literal notranslate"><span class="pre">TRUE</span></code>.  This will allow you to document an approval for example made in your CI tool and set it immediately to approved.  If you set this instead to <code class="code docutils literal notranslate"><span class="pre">FALSE</span></code>, it is possible to approve in the Merkely application.</p>
</div>
<div class="section" id="tracking-branches-for-production">
<h2>Tracking branches for production<a class="headerlink" href="#tracking-branches-for-production" title="Permalink to this headline">¶</a></h2>
<p>In a typical setup, you need to add a <code class="code docutils literal notranslate"><span class="pre">production</span></code> tracking branch in git that is updated on every deployment. For example:</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span>git checkout production
git merge --ff-only master
<span class="c1"># YOUR DEPLOYMENT COMMAND HERE</span>
git push origin production
</pre></div>
</td></tr></table></div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2021, Merkely.<br/>
    </p>
  </div>
</footer>
  </body>
</html>