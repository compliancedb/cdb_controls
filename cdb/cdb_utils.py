import getopt
import json
import os
import sys
import docker
import requests as req
from junitparser import JUnitXml
from requests.auth import HTTPBasicAuth

from cdb.settings import CDB_SERVER

CMD_HELP = 'ensure_project.py -p <project.json>'


def project_exists_in_cdb(project_data, projects):
    exists = False
    for project in projects:
        if project["name"] == project_data["name"]:
            exists = True
    return exists


def get_project_list_from_cdb(projects_url):
    resp = req.get(projects_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    projects = resp.json()
    return projects


def get_project_from_cdb(project_url):
    resp = req.get(project_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    project = resp.json()
    return project


def load_project_configuration(json_data_file):
    project_data = json.load(json_data_file)
    return project_data


def create_artifact(api_token, host, project_config_file, sha256, filename, description, git_commit, commit_url, build_url,
                    is_compliant):
    project_data = load_project_configuration(project_config_file)

    '''
    curl -H 'Content-Type: application/json' \
     -X PUT \
     -d '{"sha256": "'"$3"'", "filename": "'"$4"'", "description": "'"$5"'", "git_commit": "'"$6"'", "commit_url": "'"$7"'", "build_url": "'"$8"'", "is_compliant": "true"}' \
    http://server:8001/api/v1/projects/$1/$2/artifacts/
    '''

    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url,
        "is_compliant": is_compliant
    }
    url = url_for_artifacts(host, project_data)
    put_payload(create_artifact_payload, url, api_token)


def put_payload(payload, url, api_token):
    headers = {"Content-Type": "application/json"}
    print("Putting this payload:")
    print(json.dumps(payload, sort_keys=True, indent=4))
    print("To url: " + url)
    resp = req.put(url, data=json.dumps(payload), headers=headers, auth=HTTPBasicAuth(api_token, 'unused'))
    print(resp.text)


def url_for_artifacts(host, project_data):
    return url_for_project(host, project_data) + project_data["name"] + '/artifacts/'


def add_evidence(api_token, host, project_file_contents, sha256_digest, evidence):
    project_data = load_project_configuration(project_file_contents)
    url_for_artifact = url_for_artifacts(host, project_data) + sha256_digest

    put_payload(evidence, url_for_artifact, api_token)


def url_for_project(host, project_data):
    return host + '/api/v1/projects/' + project_data["owner"] + '/'


def rchop(thestring, ending):
    if thestring.endswith(ending):
        return thestring[:-len(ending)]
    return thestring


def parse_cmd_line():
    project_file = ""
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hp:", ["project="])
    except getopt.GetoptError:
        print(CMD_HELP)
        sys.exit(2)
    for opt, arg in opts:
        print(opt + ": " + arg)
        if opt == '-h':
            print(CMD_HELP)
            sys.exit()
        elif opt in ("-p", "--project"):
            project_file = arg
    return project_file


def get_image_details():
    docker_image = os.getenv('CDB_DOCKER_IMAGE', "NO_DOCKER_IMAGE_FOUND")
    sha256_digest = os.getenv('CDB_ARTIFACT_SHA', None)
    if sha256_digest is None:
        client = docker.from_env()
        print("Inspecting docker image for sha256Digest")
        image = client.images.get(docker_image)
        sha256_digest = image.attrs["RepoDigests"][0].split(":")[1]
    else:
        print("Docker image digest found from environment variable")

    print("IMAGE DIGEST: " + str(image.attrs["RepoDigests"][0]))
    return docker_image, sha256_digest


def env_is_compliant():
    return os.getenv('CDB_IS_COMPLIANT', "FALSE") == "TRUE"


def is_compliant_suite(junit_xml):
    if junit_xml.failures != 0:
        return False, "Tests contain failures"
    if junit_xml.errors != 0:
        return False, "Tests contain errors"
    return True, "All tests passed"


def load_test_results(file_path):
    test_xml = JUnitXml.fromfile(file_path)
    return test_xml


def is_compliant_test_results(file_path):
    """
    This parses a junit xml file to determine if there are any errors or failures

    return:
        A tuple, with is_compliant, plus a message string
    """
    test_xml = load_test_results(file_path)
    if test_xml._tag == "testsuites":
        for suite in test_xml:
            suite_is_compliant, message = is_compliant_suite(suite)
            if not suite_is_compliant:
                return suite_is_compliant, message
        # every test suite passed, so return True
        return True, "All tests passed"
    if test_xml._tag == "testsuite":
        return is_compliant_suite(test_xml)
    return False, "Could not find test suite(s)"


def control_junit():
    print("Publish evidence to ComplianceDB")
    junit_results_path = "/data/junit/junit.xml"

    is_compliant, message = is_compliant_test_results(junit_results_path)
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "junit")
    description = "JUnit results xml verified by compliancedb/cdb_controls: " + message
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def put_evidence():
    print("Publish evidence to ComplianceDB")

    is_compliant = env_is_compliant()
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "EVIDENCE_TYPE_UNDEFINED")
    description = os.getenv('CDB_DESCRIPTION', "UNDEFINED")
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def send_evidence(evidence):
    print(evidence)
    project_file = parse_cmd_line()
    with open(project_file) as project_file_contents:
        _docker_image_unused, sha256_digest = get_image_details()
        api_token = os.getenv('CDB_API_TOKEN', 'NO_API_TOKEN_DEFINED')
        host = os.getenv('CDB_HOST', CDB_SERVER)
        add_evidence(api_token, host, project_file_contents, sha256_digest, evidence)


def build_evidence_dict(is_compliant, evidence_type, description, build_url):
    evidence = {"evidence_type": evidence_type, "contents": {
        "is_compliant": is_compliant,
        "url": "",
        "description": ""
    }}
    evidence["contents"]["description"] = description
    evidence["contents"]["url"] = build_url
    return evidence
