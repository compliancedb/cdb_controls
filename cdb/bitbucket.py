import json
import os
import sys

import requests

from cdb.cdb_utils import build_evidence_dict, send_evidence


def put_bitbucket_pull_request(project_config_file):

    pull_requests = get_pull_request_for_current_commit()

    print("Found pull request info: " + str(pull_requests))
    evidence = build_evidence_dict(
        is_compliant=True,
        evidence_type="pull_request",
        description="Bitbucket pull request",
        build_url=pull_requests[0]["pullRequestURL"])
    evidence["contents"]["source"] = pull_requests
    send_evidence(project_config_file, evidence)


def get_pull_request_for_current_commit():
    workspace = os.getenv('BITBUCKET_WORKSPACE', None)
    repository = os.getenv('BITBUCKET_REPO_SLUG', None)
    commit = os.getenv('BITBUCKET_COMMIT', None)
    user = os.getenv('BITBUCKET_USER', None)
    password = os.getenv('BITBUCKET_PWD', None)
    pull_requests = get_pull_requests_from_bitbucket_api(
        workspace=workspace, repository=repository, commit=commit, username=user, password=password)
    return pull_requests


def get_pull_requests_from_bitbucket_api(workspace, repository, commit, username, password):
    pull_requests_evidence = []

    url = f"https://api.bitbucket.org/2.0/repositories/{workspace}/{repository}/commit/{commit}/pullrequests"
    r = requests.get(url, auth=(username, password))
    if r.status_code == 200:
        pull_requests_json = json.loads(r.text)
        pull_requests = pull_requests_json["values"]
        for pr in pull_requests:
            pr_evidence = {}

            pr_html_url = pr['links']['html']['href']
            pr_api_url = pr['links']['self']['href']

            pr_evidence['pullRequestMergeCommit'] = commit
            pr_evidence['pullRequestURL'] = pr_html_url

            pr_evidence = get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password)

            pull_requests_evidence.append(pr_evidence)

    elif r.status_code == 202:
        print("Repository pull requests are still being indexed, please retry.")
        sys.exit(1)
    elif r.status_code == 404:
        print("Repository does not exists or pull requests are not indexed. Please make sure Pull Request Commit Links app is installed")
        sys.exit(2)
    else:
        print(f"Exception occured in fetching pull requests. Http return code is {r.status_code}")
        sys.exit(3)

    return pull_requests_evidence


def get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password):
    r = requests.get(pr_api_url, auth=(username, password))
    if r.status_code == 200:
        pr_json = json.loads(r.text)
        pr_evidence['pullRequestState'] = pr_json['state']
        participants = pr_json['participants']
        approvers = ""
        if len(participants) > 0:
            for participant in participants:
                if participant['approved']:
                    approvers = approvers + participant['user']['display_name'] + ","
            approvers = approvers[:-1]
            pr_evidence['approvers'] = approvers
        else:
            print("No approvers found")
    else:
        print("Error occured in fetching pull request details. Please review repository permissions.")
    return pr_evidence





