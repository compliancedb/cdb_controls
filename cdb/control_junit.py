#!/usr/bin/env python

from cdb.cdb_utils import control_junit

if __name__ == '__main__':
    control_junit()
