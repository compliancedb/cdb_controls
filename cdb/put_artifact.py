#!/usr/bin/env python

import os

from cdb.cdb_utils import create_artifact, parse_cmd_line, get_image_details, env_is_compliant
from cdb.settings import CDB_SERVER


def main():
    project_file = parse_cmd_line()

    # Get the SHA and Filename from string provided in CDB_DOCKER_IMAGE
    filename = os.getenv('CDB_ARTIFACT_FILENAME', "UNDEFINED")
    if filename == "UNDEFINED":
        print("Cannot find CDB_ARTIFACT_FILENAME in the environment variables")
        return

    sha256_digest = os.getenv('CDB_ARTIFACT_SHA', "UNDEFINED")
    if sha256_digest == "UNDEFINED":
        print("Cannot find CDB_ARTIFACT_SHA in the environment variables")
        return

    print("Publish artifact to ComplianceDB")
    host = os.getenv('CDB_HOST', CDB_SERVER)

    description = "Created by build " + os.getenv('CDB_BUILD_NUMBER', "UNDEFINED")
    git_commit = os.getenv('CDB_ARTIFACT_GIT_COMMIT', '0000000000000000000000000000000000000000')

    commit_url = os.getenv('CDB_ARTIFACT_GIT_URL', "GIT_URL_UNDEFINED")
    build_url = os.getenv('CDB_CI_BUILD_URL', "BUILD_URL_UNDEFINED")
    is_compliant = env_is_compliant()

    print('CDB_IS_COMPLIANT: ' + str(is_compliant))

    api_token = os.getenv('CDB_API_TOKEN', 'NO_API_TOKEN_DEFINED')

    with open(project_file) as project_file_contents:
        create_artifact(api_token, host, project_file_contents, sha256_digest, filename, description, git_commit,
                        commit_url, build_url, is_compliant)


if __name__ == '__main__':
    main()
