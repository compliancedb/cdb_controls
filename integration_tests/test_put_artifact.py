from cdb.put_artifact import put_artifact

from os import environ
from approvaltests.approvals import verify
from approvaltests.reporters import PythonNativeReporter
from tests.cdb_dry_run import cdb_dry_run


def test_message_when_env_var_CDB_ARTIFACT_FILENAME_is_missing(capsys):
    with cdb_dry_run():
        put_artifact("integration_tests/test-pipefile.json")

    captured = capsys.readouterr()
    verify(captured.out, PythonNativeReporter())


def test_message_when_env_var_CDB_ARTIFACT_SHA_is_missing(capsys):
    with cdb_dry_run():
        with SetEnv({"CDB_ARTIFACT_FILENAME": "tests_data/put-artifact.txt"}):
            with SetEnv({"CDB_ARTIFACT_SHA": "UNDEFINED"}):
                put_artifact("integration_tests/test-pipefile.json")

    captured = capsys.readouterr()
    verify(captured.out, PythonNativeReporter())


class SetEnv(object):
    def __init__(self, env_vars):
        self._env_vars = env_vars
        for (name, value) in env_vars.items():
            environ[name] = value

    def __enter__(self):
        return self

    def __exit__(self, _type, _value, _traceback):
        for name in self._env_vars:
            environ.pop(name)


