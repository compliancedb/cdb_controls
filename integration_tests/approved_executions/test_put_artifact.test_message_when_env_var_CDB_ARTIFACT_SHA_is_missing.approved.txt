Cannot find CDB_ARTIFACT_SHA in the environment variables
